### ISSUE TITLE
How to iterate over member array in global memory?

### ISSUE THREAD

**Issue by devreal** [dash-project/dash#586]
Consider the following example:

```C
struct data_t {
  int a, b;
  int c[NUM_VALS];
};

dash::Shared<data_t> data(data_t{});
```

I can access individual members of `data_t` using the `member()` function of `GlobRef`:
```C
dash::GlobRef<int> a_ref = data.get().member(&data_t::a);
```

However, I cannot figure out a way to access individual elements of the member array `c`. I can get a reference to the first element (using `member(&data_t::c)`) but I cannot create a global pointer from that reference (as I am missing the memory space). 

Any ideas? This might be a use-case to consider while working on the memory spaces @rkowalewski?

**Comment by rkowalewski** [comment]
One use case I recently had is a `dash::Shared<int[N]>`. See this [implementation](https://github.com/dash-project/dash-apps/blob/c147ad489abe315fb411e4ec45b8512a1ae2fdac/splash2/barnes-hut/shared_array.h) as an example. This eventually helps you for your use case.
A generic solution basically requires something like a constexpr array which can give you the relevant offsets at compile time. However, if we start with something like this then several questions come up:

- What about multiple dimensions?
- In case of multiple dimensions, what about memory layout (row major vs column major)?

IMHO it adds complexity and additional burden of maintenance. I would suggest implement what you need for the special case. And think about alternative solutions, maybe better and more simple.

**Comment by devreal** [comment]
> - What about multiple dimensions?
> - In case of multiple dimensions, what about memory layout (row major vs column major)?

No no no. The offset into an array member of a struct in the global address space is left to the user and nothing we care about (it is row-major, just as the C++ standard defines it). It should not be so hard to do this and certainly not require the user to write his own abstraction (WTF?!)

All that is needed is a way to request a `GlobPtr` from a `GlobRef` that can then be used to iterate over elements, just like you get an `int*` from a `int&`. With this you can iterate over the range of the member array, easy. The problem is the current design of `GlobPtr`, which has the underlying memory space typed in (WTF^2?!)

There have been discussions on this in the past and we definitely have to discuss this again during the F2F. From my point of view, there are three main requirements for GlobPtr:

1. Its type is only dependent on the underlying value type.
2. It is `trivially_copyable` so we can move it around across unit boundaries.
3. It is constructible from a `GlobRef` alone. 

(Please amend the list as you see fit.)

A simple implementation has the following fields:

```C
template<typename T>
class GlobPtr {
  size_t lb, ub; // lower and upper bound of the current range 
  dart_gptr_t gptr; // contains teamid, segmentid, offset
};
```

The memory space information is abstracted away into a (process-)global `unordered_list` that is indexed by the `{teamid, segmentid}` pair that can be drawn from the `gptr`, which is globally unique and available on all units in the team. The fields `lb` and `ub` are globally valid and define the current bounds of `gptr.addr_or_offset.offset`. Only if an increment/decrement on the `GlobPtr` causes `gptr.addr_or_offset.offset` to cross these range boundaries we have to query the `unordered_list` to jump to the next range and update these fields using the memory space information. This can be done on any unit and in `O(1)`.

In this scheme, memory spaces are abstracted behind an interface and implemented in terms of virtual inheritance but the overhead should really be negligible as access to the underlying memory space is rather rare (after all, `GlobPtr` increments are fine as long as they don't cross range boundaries).

AFAICS, this fulfills all three requirements outlined above. Please correct me if I missed something there and we can incorporate any missing information into this scheme. I'm happy to provide a PoC of this if necessary.

**Comment by rkowalewski** [comment]
Keep Calm and drink some beer. Your response goes way beyond the scope of your initial question. ;) It is actually a completely different issue as accessing an array at a particular index is not the same as iterating raw memory, in the strict sense and from the implementation point of view.

> There have been discussions on this in the past and we definitely have to discuss this again during the F2F.

Yup, most stuff what you mention above is done, so we can do a hands-on session.

**Comment by rkowalewski** [comment]
> I'm happy to provide a PoC of this if necessary.

We do not need any PoC. DASH is full of it. We need a working and well tested library.

**Comment by devreal** [comment]
> Keep Calm and drink some beer. Your response goes way beyond the scope of your initial question. ;)

I agree. But I cannot accept that this should not be easily doable and would require me to implement a hack just because our implementation is bad (and I have criticized this several times before and not seen a solution...)

> Yup, most stuff what you mention above is done, so we can do a hands-on session.

Looking forward to that :)

**Comment by rkowalewski** [comment]
> I agree. But I cannot accept that this should not be easily doable and would require me to implement a hack just because our implementation is bad (and I have criticized this several times before and not seen a solution...)

You first talked about `dash::Shared<int[N>` and how to access individual elements. Now we are talking about simple pointer arithmetic. These are two different things at least, apparently. For your first question I provided you a prototypical solution, eventually. Regarding your memory thing, let us make an agreement and fix conceptual inconsistencies in Stuttgart. The only thing I want from you: Make sure that we have some beer available.

**Comment by devreal** [comment]
Will be possible with `dash::addressof`, see #588

**Comment by rkowalewski** [comment]
fixed in #515
